### PR TITLE
MWPW-135790 Optimize migration

### PR DESCRIPTION
**Pull request by Brandon32**
* Use callback for loading markdown files
* Skip pages with no blocks to migrate
* Cache docx files for faster migrations

Resolves: [MWPW-135790](https://jira.corp.adobe.com/browse/MWPW-135790)

### CODE REVIEW

File: block-report.js [hunk 1]
```diff
@@ -14,113 +14,123 @@ import xlsx from 'xlsx';
...
         const mdast = await getMdast(markdown);
         const tableMap = getTableMap(mdast);
 
         for (let j = 0; j < tableMap.length; j++) {
+            const blockReport = { links: [] };
+
             const table = tableMap[j];
             const { blockName, options } = table;
+            blockReport.name = blockName;
+            blockReport.index = j;
```
Comment (meganthecoder) [inline block-report.js:47]: Nit: combine these
```suggestion
            const table = tableMap[j];
            const { blockName, options } = table;
            const blockReport = {
              name: blockName,
              index: j,
              links: [],
            };
```

File: block-report.js [hunk 2]
```diff
@@ -14,113 +14,123 @@ import xlsx from 'xlsx';
...
-                blockLinks[name].push(url);
+            links.forEach(({ url }) => {
+                blockReport.links.push(url);
             });
+            entryReport.blocks.push(blockReport);
         };
 
         const links = getNodesByType(mdast, 'link');
         if (links) {
             links.forEach((link) => {
```
Comment (meganthecoder) [inline block-report.js:64]: Nit 
```suggestion
            links?.forEach((link) => {
```

File: block-report.js [hunk 3]
```diff
@@ -14,113 +14,123 @@ import xlsx from 'xlsx';
...
+    // Block Links
+    console.log('creating block links sheet');
     const blockLinksData = [['Path', 'URL', 'Block', 'Index', 'Links']];
-    for (const entry in blockLinks) {
-        const [entryName, blockName, index] = entry.split(',');
-        const row = [entryName, `${site}${entryName}`, blockName, index, ...blockLinks[entry]];
-        blockLinksData.push(row);
+    for (const entry in report) {
+        report[entry].blocks.map((block) => {
+            if(block.links.length === 0) return;
```
Comment (meganthecoder) [inline block-report.js:124]:
```suggestion
            if (block.links.length === 0) return;
```

File: blog-migration.js [hunk 4]
```diff
@@ -10,78 +10,109 @@
...
+            const sourceFolder = `${docxDir}${path}`;
+            const outputFolder = `${outputDir}${path}`
+            const sourceDocxFile = `${sourceFolder}/${docxFile}`;
+            const outputDocxFile = `${outputFolder}/${docxFile}`;
+
             const mdast = await getMdast(markdown);
+            const tableMap = getTableMap(mdast);
+            if (!tableMap.reduce((acc, block) => {
+                return acc || Object.keys(MIGRATION).includes(block.blockName);
+            }, false)) {
```
Comment (meganthecoder) [inline blog-migration.js:64]: Nit: This is hard to read in my opinion. Might want to consider breaking it out

File: utils/docx-utils.js [hunk 5]
```diff
@@ -1,31 +1,217 @@
...
+
+    try {
+        await readFile(outputFile);
+        console.log(`Skipping ${outputFile} as docx already exists.`);
+        return;
+    } catch (err) {
+        const buffer = await updateDocx(mdast, sourceFile);
+        await writeFile(outputFile, buffer);
+        return;
+    }
```
Comment (meganthecoder) [inline utils/docx-utils.js:76]: Nit: unnecessary
```suggestion
    try {
        await readFile(outputFile);
        console.log(`Skipping ${outputFile} as docx already exists.`);
    } catch (err) {
        const buffer = await updateDocx(mdast, sourceFile);
        await writeFile(outputFile, buffer);
    }
```

**Review by meganthecoder — APPROVED** [review]
Looks good. Just a few nits

File: blog-migration.js (hunk 4 above)
Comment (Brandon32) [reply blog-migration.js:64]: Thanks, using `tableMap.some` would make this more clear.

File: fetch-markdown.js [hunk 6]
```diff
@@ -26,33 +26,31 @@ const REPORT_DIR = 'reports';
...
+    const totals = { succeed: 0, failed: 0 };
+    const failed = [];
     const entries = await loadIndex(project, indexUrl, cached);
+    const mdFolder = `./${MD_DIR}/${project}`;
 
-    for (let i = 0; i < entries.length; i++) {
-        const entry = entries[i];
-        const markdown = await loadMarkdown(`${site}${entry}`, `./${MD_DIR}/${project}${entry}`, cached);
-
+    await loadMarkdowns(site, mdFolder, entries, cached, (markdown, entry, i) => {
```
Comment (JasonHowellSlavin) [inline fetch-markdown.js:34]: We might want to discuss how this is built at some point. 5 params, one of which is a callback, is a lot of params.

File: block-report.js [hunk 7]
```diff
@@ -14,118 +14,127 @@ import xlsx from 'xlsx';
...
     const allBlocks = Object.keys(totals.blocks).sort();
     const ws_data = [['Path', 'URL', ...allBlocks]];
 
-    for (const entry in blocks) {
-        const blockValues = allBlocks.map((key) => blocks[entry][key] || 0);
-        const row = [entry, `${site}${entry}`, ...blockValues];
+    // Block Entries
+    console.log('creating block entries sheet');
+    for (const entry in report) {
+        const blocks = allBlocks.map((key) => report[entry].blocks.reduce((acc, block) => {
```
Comment (JasonHowellSlavin) [inline block-report.js:82]: block level function calls here might add some readability, if you thought it prudent. Would add a few lines with `reutrn`s though.

**Review by JasonHowellSlavin — APPROVED** [review]
Is anything in docx utils straight from helix code? If so we may run into issues with the public git? What do you think? 

Looks good, few small items

File: fetch-markdown.js (hunk 6 above)
Comment (Brandon32) [reply fetch-markdown.js:34]: Having more than 3 parameter can be sign that a function is doing too much but also can be unavoidable or even preferable. 
`loadMarkdowns` calls `loadMarkdown` so I need 3 parameters plus the callback. We could to reduce parameter count by pre-mapping the entries into a array of objects, `[{mdUrl, mdPath}, ...]`, but now we are passing a lot of redundant data with similar complexity.

We can move the callback to a named function, and that seems like a good idea.
Also we can have a separate functions for cached and uncached load markdowns, boolean parameter suggest this.
```js
loadMarkdowns(entries, site, callback)
loadMarkdownsCached(entries, mdFolder, site, callback)
```
Not sure how useful that is yet, we need to use this first.

File: block-report.js (hunk 7 above)
Comment (Brandon32) [reply block-report.js:82]: Maybe after we run some more reports we can refactor this.

**Pull request merged**